README.md

```diff
@@ -0,0 +1,109 @@
+   * Merging structured data into even bigger trees helped to keep all data in easy to edit, human readable files, even with the option to pull in 'override files' to patch API descriptions as needed. The latter was used with the [Drive API](https://github.com/Byron/google-apis-rs/blob/master/etc/api/drive/v2/drive-api_overrides.yaml#L1), even though I would only add such capability on an as-needed basis.
+1. **Having an off-the-shelf template engine**
+   * Generating text from 'smart' files with some basic support for syntax highlighting made it quite easy to get started.
+   * This was only possible through the `make` driven automation, as one would have to run multiple tools including `cargo check` to
+    see if it actually worked.
+1. **Having as much code as possible available without template**
+   * particularly including `util.rs` helped to provide common code, and I would put as much code as possible into Rust-only files.
+   * code size could further be reduced by putting that code into its own crate and maintain it as such.
+1. **Performance**
+   * Code generation was fast enough and could be parallelized on a per-API/CLI basis thanks to `make`.
+1. **API, Docs, and CLIs**
+   * I think having fully type-safe and chainable APIs was great.
+   * The docs were lovely
+   * The CLIs allowed to use an API almost instantly
+
+## 🥵Issues with _OP_'s way of doing things🥵
+
+1. **logic in templates**
+   * What seemed like a huge benefit was also causing vastly difficult to read and understand templates.
+   * I remember that sometimes, I worked around certain limitations of the engine, which masked what one was actually doing.
+   * Separation of concerns is a good thing, but _OP_ didn't have it. The template engine transformed the data model from _discovery_
+    on the fly.
+1. **massive code duplication caused huge libraries**
+   * Each call would be 'spelled out', and APIs came with many calls. This caused massive libraries that take a while to check and build.
+   * Huge files are harder to read
+1. **improper namespace handling**
+   * types introduced by the API could clash with reserved names of the Rust language, or with names imported into the namespace.
+   * It wasn't easy to handle names consistently in all places that needed them
+1. **python and mako**
+   * Even though they worked, it was another thing that had to be installed by `make`, and could just fail [for some](https://github.com/Byron/google-apis-rs/issues/234)
+1. **arbitrary smartness**
+   * In order to fix issues with the type system and make numbers more easily usable by converting "strings" into integers/floats, what worked in one API broke another.
+1. **it's cumbersome to actually use a CLI**
+   * Even though authentication was dealt with nicely for the most part, actually using APIs required them to be enabled via the developer console. From there one would download a file and deposit it in the right spot. Only then one could start using the CLI.
```

Comment:
I haven't used the CLI's, but I'm curious to hear more about this. Couldn't a single google developer project be created for this purpose? It would enable all the API's and credentials could be created that get bundled with the CLI? I'm probably misunderstanding the issue or missing something obvious.

Reply:
No, you are getting it absolutely right. In the first google apis release, I did exactly that! However, it was incredibly flaky and the credentials stopped working quickly.
Thus I would rather try to make it very easy for everyone to setup (and manage) their own API allowances.

The CLIs were designed to expose the entire capability of the API, a true power tool. I would want that again, but with more convenience when getting started. I saw myself being hesitant to try new APIs because of the overhead.